### PR TITLE
added a sanitize_filename method

### PR DESCRIPTION
**Pull request by ThunderKey**

### CODE REVIEW

**Review by The-Compiler — CHANGES_REQUESTED** [review]
This will break things, because `syncing.py` tries then to ask for the filenames the plugin returned, and those don't actually exist on the server.

Instead, `syncing.py` should probably only sanitize the output filename, no?

**Comment by ThunderKey** [comment]
Oh yeah...

The problem would be Folders with slashes in them. This would lead to sub folders, which could be confusing. E.g. "17./18. of March"

**Comment by ThunderKey** [comment]
But I guess it's the best option. The only alternative I'd see would be a "translation" dictionary, which would make everything much more complicated

File: src/kitovu/sync/syncing.py [hunk 1]
```diff
@@ -77,7 +77,8 @@ def start(connection_settings: ConnectionSettings, reporter: utils.AbstractRepor
             # remote_full_path: /Informatik/Fachbereich/EPJ/Dokumente/Anleitung.pdf
             #   with relative_to: Dokumente/Anleitung.pdf
             # -> local_full_path: /home/leonie/HSR/EPJ/Dokumente/Anleitung.pdf
-            local_full_path: pathlib.Path = local_dir / remote_full_path.relative_to(remote_dir)
+            local_full_path: pathlib.Path = local_dir / \
+                utils.sanitize_filename(str(remote_full_path.relative_to(remote_dir)))
```
Comment (The-Compiler) [inline src/kitovu/sync/syncing.py]: I'd prefer to add a second variable to break this, e.g.

```python
filename: str = utils.sanitize_filename(str(remote_full_path.relative_to(remote_dir)))
local_full_path: pathlib.Path = local_dir / filename
```

File: src/kitovu/utils.py [hunk 2]
```diff
@@ -24,6 +24,26 @@ def get_password(plugin: str, identifier: str) -> str:
     return password
 
 
+def sanitize_filename(name: str, replacement: typing.Optional[str] = '_') -> str:
```
Comment (The-Compiler) [inline src/kitovu/utils.py]: Might be more in line with what we do elsewhere to take a `pathlib.PurePath` (and return one) here instead.

File: src/kitovu/utils.py [hunk 3]
```diff
@@ -24,6 +24,26 @@ def get_password(plugin: str, identifier: str) -> str:
...
+    r"""Replace invalid filename characters.
+
+    Note: This does not escape directory separators (/ and \).
+
+    Args:
+        name: The filename.
+        replacement: The replacement character (or None).
+    """
```
Comment (The-Compiler) [inline src/kitovu/utils.py]: Probably okay to drop at least the `Args:` part of the docstring here.

File: src/kitovu/utils.py [hunk 4]
```diff
@@ -24,6 +24,26 @@ def get_password(plugin: str, identifier: str) -> str:
...
+    if replacement is None:
+        replacement = ''
```
Comment (The-Compiler) [inline src/kitovu/utils.py]: We don't need this for kitovu, so let's remove it. Actually, we don't need the ability to customize the replacement char at all.

File: tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import pytest
+
+from kitovu import utils
+
+
+class TestSanitizeFilename():
```
Comment (The-Compiler) [inline tests/test_utils.py]: Drop the `()` for consistency with other classes.

File: tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
+import pytest
```
Comment (The-Compiler) [inline tests/test_utils.py:1]: Also see https://github.com/qutebrowser/qutebrowser/blob/v1.3.0/tests/unit/utils/test_utils.py#L632-L638 for my tests, but I think yours are fine as well.

**Review by The-Compiler — CHANGES_REQUESTED** [review]

**Review by The-Compiler — APPROVED** [review]

**Comment by The-Compiler** [comment]
I opened [EPJ-94](https://jira.keltec.ch/jira/browse/EPJ-94) for how to handle slashes in filenames. Like you say, it'd be nice to handle this in a nicer way.

**Pull request merged**